### ISSUE TITLE
Undefined TermPos when processing typed block

### ISSUE THREAD

**Issue by vkleen** [tweag/nickel#1226]
The following code yields a `TermPos::unwrap` panic at `src/label.rs:248:35`.
```nickel
{
  Contract | {},
  Config = {
    x | Contract = {},
    y = lib.f x,
  },
  lib = {
    f | {_: Dyn} -> {_: Dyn} = fun x => x,
  }
} :_
```

**Comment by vkleen** [comment]
But it happens only sometimes, when run with `nickel -f repro.ncl`. Presumably, that's the randomized evaluation order caused by differen `HashMap` seeds.

**Comment by yannham** [comment]
Are you using an up-to-date master branch? I don't see an unwrap at line `248`, but I see one at line `245`, which I'm hopefully fixing right now as part of the merging arrays PR. At least some of them.

**Comment by yannham** [comment]
By the way, did you mean `Contract = {}` instead of `Contract | {}` ?

**Comment by yannham** [comment]
The issue is that some static type errors, most notably `ArrowTypeMismatch`, have been using the same infrastructure as contract's type paths to report nice underlining of subtype in a more complex type. When the type was pretty-printed and the offset computed, everything was fine. But then we added position to types and said oh we want to show the subtype _in situ_, where the contract has been defined.

This works well for contract, but the static types are (most of them) generated by type inference, so they don't have a position, and everything is blowing up. My current fix will be to make the function fallible, and if we found that the subtype has no position, we pretty-print the type in a temporary source (a bit like what we do for some terms, the infamous `<unknown> (generated by evaluation)`, but for types), and do everything one more time. Now the type and all of its subtype have a position, and hence should be fine
